File: recipe-0005/cxx-example/geometry_polygon.cpp
#include "geometry_polygon.hpp"

#include <cmath>

namespace geometry {
namespace area {
double polygon(int nSides, double side) {
  double perimeter = nSides * side;
  double apothem = side / (2.0 * std::tan(M_PI / nSides));
  return (perimeter * apothem) / 2.0;
}
}
}

File: chapter-01/recipe-09/fortran-example/geometry_polygon.f90
module class_Polygon

  implicit none
  private
  real :: pi = acos(-1.0d0)

  type, public :: Polygon
    integer :: nSides
    real :: side
  contains
    procedure :: area => polygon_area
    procedure :: print => polygon_print
  end type

contains

  pure function polygon_area(this) result(area)

    class(Polygon), intent(in) :: this
    real :: area
    real :: perimeter, apothem

    perimeter = this%nSides * this%side
    apothem = this%side / (2.0d0 * tan(pi / this%nSides))

    area = (perimeter * apothem) / 2.0d0

  end function

  subroutine polygon_print(this)

    class(Polygon), intent(in) :: this
    real:: area

    area = this%area()
    print *, 'Polygon: nSides = ', this%nSides, ' side = ', this%side, ' area = ', area

  end subroutine

end module
